src/secdb.c

```diff
@@ -0,0 +1,847 @@
+					 &sig, &sigsz);
+		if (rc < 0 && errno == EOVERFLOW) {
+			debug("esl_iter_next at %zd(0x%zx) is malformed; attempting correction",
+			      esd_get_esl_offset(iter), esd_get_esl_offset(iter));
+			corrected = true;
+			rc = esl_iter_next_with_size_correction(iter,
+					&secdb_type_guid, &owner, &sig, &sigsz,
+					true);
+			debug("got new entry at 0x%zx with sigsz:%zd",
+			      esd_get_esl_offset(iter), sigsz);
+		}
+		if (rc < 0) {
+			efi_error("Could not get next security database entry");
+			esl_iter_end(iter);
+			if (new_secdb)
+				free(secdb);
+			return rc;
+		}
+		if (rc == ESL_ITER_DONE)
+			break;
+
+		if (new_secdb)
+			secdb->sigsz = sigsz;
+		debug("sigsz:%zd", sigsz);
+                secdb_type = secdb_entry_type_from_guid(&secdb_type_guid);
+		debug("secdb_type:%d", secdb_type);
+
+		if (corrected)
+			force = true;
+		if (rc == ESL_ITER_NEW_LIST && !sort)
+			force = true;
+		if (new_secdb)
+			force = false;
+
+		if (force) {
+			if (corrected)
+				debug("forcing new secdb due to size correction");
+			else if (rc == ESL_ITER_NEW_LIST && !sort)
+				debug("forcing new secdb due to new input ESL sort!=type");
+			else
+				debug("wth?  why is force set");
+		}
+
+		efi_secdb_add_entry_or_secdb(secdb, &owner, secdb_type,
+					     (efi_secdb_data_t *)sig, sigsz,
+					     force);
+		new_secdb = false;
+	} while (rc > 0);
+
+	esl_iter_end(iter);
+
+	if (sort) {
+		debug("sorting lists %s", sort_descending ? "desc" : "asc");
+		int (*cmp)(const void *, const void *, void *) =
+			(sort_descending
+			 ? secdb_cmp_descending
+			 : secdb_cmp);
+		list_sort(&secdb->list, cmp, NULL);
+	}
+
+	*secdbp = secdb;
+	return 0;
+}
+
+struct visitor_state {
+	/* listnum from the previous invocation */
+	unsigned int listnum;
+
+	efi_signature_list_t *esl;
+
+	char *buf;
+	size_t pos;
+};
+
+/*
+ * realize a signature list file from our internal representation into
+ */
+static efi_secdb_visitor_status_t
+secdb_realize_visitor(unsigned int listnum, unsigned int signum,
+		      const efi_guid_t * const owner,
+		      const efi_secdb_type_t algorithm,
+		      const void * const header,
+		      const size_t headersz,
+		      const efi_secdb_data_t * const data,
+		      const size_t datasz,
+		      void *closure)
+{
+	struct visitor_state *state = closure;
+	const efi_guid_t *alg = secdb_guid_from_type(algorithm);
+	char *buf;
+	size_t allocsz, esdsz;
+	ptrdiff_t skew;
+	efi_signature_list_t *esl;
+	efi_signature_data_t *esd;
+	bool has_owner = true;
+	int rc;
+
+	rc = secdb_entry_has_owner_from_type(algorithm, &has_owner);
+	if (rc < 0)
+		efi_error("could not determine signature type");
+
+	esdsz = datasz + (has_owner ? sizeof(efi_guid_t) : 0);
+
+	debug("listnum:%d signum:%d has_owner:%d", listnum, signum, has_owner);
+	if (listnum > state->listnum || signum == 0) {
+		allocsz = ALIGN_UP(state->pos
+				   + sizeof(state->esl)
+				   + headersz
+				   + esdsz,
+				   page_size);
```

Comment:
Don't think I agree with breaking this.  It only needs to be two lines:

```C
    allocsz = ALIGN_UP(state->pos + sizeof(state->esl) + headersz + esdsz +
                       page_size);
```

Reply:
(Ignoring your + vs , typo)
If I do that, you're going to ask me to re-wrap it because "esdsz," ends at column 87, aren't you?  I think instead I'm just going to do:
```
                allocsz = state->pos + sizeof(state->esl) + headersz + esdsz;
                allocsz = ALIGN_UP(allocsz, page_size);
```